control-plane/agents/src/bin/ha/node/server.rs

```diff
@@ -239,13 +257,38 @@ impl NodeAgentOperations for NodeAgentSvc {
         // path has been successfully created, so having the first failed path in addition
         // to the second healthy one is OK: just display a warning and proceed as if
         // the call has completed successfully.
-        disconnect_controller(&ctrlr).or_else(|e| {
+        if let Err(error) = disconnect_controller(&ctrlr, request.new_path()) {
             tracing::warn!(
                 uri=%request.new_path(),
-                error=%e,
+                error=%error,
                 "Failed to disconnect failed path"
             );
-            Ok(())
-        })
+        };
+        Ok(())
+    }
+}
+
+// Returns the host, port, nqn respectively from the path after parsing.
+fn parse_uri(new_path: &str) -> Option<(String, String, String)> {
+    let uri = new_path.parse::<Uri>().ok()?;
+
+    let host = uri.host()?;
+    let port = uri.port()?.to_string();
+    let nqn = &uri.path()[1 ..];
+    Some((host.to_string(), port, nqn.to_string()))
+}
+
+// Returns the particular subsystem based on the nqn and address.
+fn get_subsystem(parsed_uri: (String, String, String)) -> Result<Subsystem, SvcError> {
```

Comment:
having an Uri instead of an unnamed tuple is cleaner here IMHO.
another alternative is to create our own parsed uri type with pub getters.

Reply:
Going it `ParsedUri` type as that suits this use case.